mkosi

```diff
@@ -2204,42 +2250,114 @@ def install_arch(args: CommandLineArguments, root: str, do_run_build_script: boo
-        ]
+        os.makedirs(hooks_dir, 0o755, exist_ok=True)
+        os.makedirs(scripts_dir, 0o755, exist_ok=True)
 
-        try:
-            return run(cmdline + args, check=True)
-        finally:
-            stop_gpg_agent()
+        # Disable depmod pacman hook as depmod is handled by kernel-install as well.
+        os.symlink("/dev/null", os.path.join(hooks_dir, "60-depmod.hook"))
+
+        kernel_add_hook = os.path.join(hooks_dir, "kernel-add.hook")
+        with open(kernel_add_hook, 'w') as f:
+            f.write(
+                dedent(
+                    """\
+                    [Trigger]
+                    Operation = Install
+                    Operation = Upgrade
+                    Type = Path
+                    Target = usr/lib/modules/*/vmlinuz
+
+                    [Trigger]
+                    Operation = Install
+                    Operation = Upgrade
+                    Operation = Remove
+                    Type = Package
+                    Target = systemd
+                    Target = dracut
+                    Target = intel-ucode
+                    Target = amd-ucode
+
+                    [Action]
+                    Description = Adding kernel and initramfs images to /boot...
+                    When = PostTransaction
+                    Exec = /etc/pacman.d/scripts/kernel-add
+                    NeedsTargets
+                    """
+                )
+            )
+
+        kernel_add_script = os.path.join(scripts_dir, "kernel-add")
+        with open(kernel_add_script, 'w') as f:
+            f.write(
+                dedent(
+                    """\
+                    #!/bin/bash -e
+
+                    ALL=false
+
+                    read -r -d '' -a TARGETS || true
+                    for i in "${!TARGETS[@]}"; do
+                        [[ ${TARGETS[$i]} != */vmlinuz ]] && ALL=true
+                        TARGETS[$i]="/${TARGETS[$i]}"
+                    done
+
+                    [[ "$ALL" = true ]] && TARGETS=(/usr/lib/modules/*/vmlinuz)
+
+                    for f in "${TARGETS[@]}"; do
+                        kernel-install add $(basename $(dirname $f)) $f
+                    done
+                    """
+                )
+            )
+
+        make_executable(kernel_add_script)
+
+        kernel_remove_hook = os.path.join(hooks_dir, "kernel-remove.hook")
+        with open(kernel_remove_hook, 'w') as f:
+            f.write(
+                dedent(
+                    """\
+                    [Trigger]
+                    Operation = Upgrade
+                    Operation = Remove
+                    Type = File
+                    Target = usr/lib/modules/*/vmlinuz
+
+                    [Action]
+                    Description = Removing kernel and initramfs images from /boot...
+                    When = PreTransaction
+                    Exec = /bin/bash -c 'while read -r f; do kernel-install remove $(basename $(dirname $f)); done'
```

Comment:
Shouldn't this call `kernel_remove_script`?

Reply:
Ouch! Good catch!